bot/utils/redis_dict.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+import json
+from collections.abc import MutableMapping
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import redis as redis_py
+
+from bot import constants
+
+ValidRedisKey = Union[str, int, float]
+JSONSerializableType = Optional[Union[str, float, bool, Dict, List, Tuple, Enum]]
+
+
+class RedisDict(MutableMapping):
+    """
+    A dictionary interface for a Redis database.
+
+    Objects created by this class should mostly behave like a normal dictionary,
+    but will store all the data in our Redis database for persistence between restarts.
+
+    Redis is limited to simple types, so to allow you to store collections like lists
+    and dictionaries, we JSON deserialize every value. That means that it will not be possible
+    to store complex objects, only stuff like strings, numbers, and collections of strings and numbers.
+    """
+
+    _namespaces = []
+    _redis = redis_py.Redis(
```

Comment:
This won't work in production without a redis password. I have put the actual variable name in the org issue iirc but from memory it is REDIS_PASSWORD.

Reply:
Fixed in ecf7f24